<a href="https://colab.research.google.com/github/vladimiralencar/DeepLearning-LANA/blob/master/DNN/DNN_Keras_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Deep Neural Networks com TensorFlow 
## Classificação

![TensorFlow](http://localhost:8888/notebooks/Dropbox/BigDataAnalytics/DeepLearning/DSA/FIA/deepLearning_I/cap06-DeepNeuralNetwork/Exercicio8/images/tensorflow.png "TensorFlow")

TensorFlow é uma biblioteca de software de código aberto para aprendizagem de máquinas em vários tipos de tarefas de compreensão de percepção e linguagem. Atualmente, ele é usado tanto para pesquisa quanto para produção por diferentes equipes em muitos produtos comerciais do Google, como reconhecimento de fala, Gmail, Google Photos e pesquisas, muitas das quais anteriormente utilizaram seu antecessor DistBelief. O TensorFlow foi originalmente desenvolvido pela equipe do Google Brain para fins de pesquisa e produção do Google e posteriormente lançado sob a licença de código aberto Apache 2.0 em 9 de novembro de 2015.

* [TensorFlow Homepage](https://www.tensorflow.org/)
* [TensorFlow GitHib](https://github.com/tensorflow/tensorflow)
* [TensorFlow Google Groups Support](https://groups.google.com/forum/#!forum/tensorflow)
* [TensorFlow Google Groups Developer Discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/discuss)
* [TensorFlow FAQ](https://www.tensorflow.org/resources/faq)


In [1]:
import tensorflow as tf
print("Versão do TensorFlow: {}".format(tf.__version__))

Versão do TensorFlow: 1.12.0


# Usando TensorFlow

TensorFlow é uma API de matemática de baixo nível, semelhante ao [Numpy] (http://www.numpy.org/). No entanto, ao contrário de Numpy, TensorFlow é construído para aprendizagem profunda. O TensorFlow funciona permitindo que você defina grafos de computação com o Python. O TensorFlow compila esses grafos de computação em um código C ++ / [CUDA] (https://developer.nvidia.com/cuda-zone) altamente eficiente.

Neste exercício, seu trabalho é criar as camadas da rede neural, necessárias para construir o modelo de classificação, usando TensorFlow e Keras. Usaremos o dataset iris.csv e nosso objetivo é prever a categoria (ou classe) a qual pertencem as plantas. 

### Funções de Transformação (função de apoio para construção do modelo)

In [0]:
# Encode dos valores de texto (i.e. [1],[2],[3] para vermelho, verde e azul).
from sklearn import preprocessing
def encode_text_index(df, name):
    le = preprocessing.LabelEncoder()
    df[name] = le.fit_transform(df[name])
    return le.classes_

# Função para preencher os valores NA
def missing_median(df, name):
    med = df[name].median()
    df[name] = df[name].fillna(med)
    
# Converte um dataframe do Pandas para inputs x,y que TensorFlow precisa
def to_xy(df, target):
    result = []
    for x in df.columns:
        if x != target:
            result.append(x)
    
    # Descubrindo o tipo da coluna de destino.  
    target_type = df[target].dtypes
    target_type = target_type[0] if hasattr(target_type, '__iter__') else target_type
    
    # Codificação. TensorFlow gosta de 32 bits.
    if target_type in (np.int64, np.int32):
        # Classification
        dummies = pd.get_dummies(df[target])
        return df.as_matrix(result).astype(np.float32), dummies.as_matrix().astype(np.float32)
    else:
        # Regression
        return df.as_matrix(result).astype(np.float32), df.as_matrix([target]).astype(np.float32)

### Carregando os dados

In [3]:
from google.colab import files
files.upload()
!mkdir data
!cp *.csv data

Saving iris.csv to iris.csv


In [4]:
!ls data

iris.csv


In [5]:
import pandas as pd
import io
import os
import requests
import numpy as np
from sklearn import metrics
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.callbacks import EarlyStopping
path = "./data"

# Carrega o dataset
filename_read = os.path.join(path,"iris.csv")

# Preenche com o valor NA quando não houver dados na coluna
df = pd.read_csv(filename_read, na_values = ['NA','?'])

# Encode das classes
species = encode_text_index(df,"species")

# Converte para o formato x,y requerido pelo TensorFlow
x,y = to_xy(df,"species")



Using TensorFlow backend.


In [6]:
df.head()

sepal_l  sepal_w  petal_l  petal_w  species
0      5.1      3.5      1.4      0.2        0
1      4.9      3.0      1.4      0.2        0
2      4.7      3.2      1.3      0.2        0
3      4.6      3.1      1.5      0.2        0
4      5.0      3.6      1.4      0.2        0

In [9]:
x.shape

(150, 4)

In [10]:
y.shape

(150, 3)

In [8]:
# Cria a rede neural
model = Sequential()
model.add(Dense(10, input_dim = x.shape[1], kernel_initializer = 'normal', activation = 'relu'))
model.add(Dense(1, kernel_initializer = 'normal'))
model.add(Dense(y.shape[1], activation = 'softmax' ))
model.compile(loss = 'categorical_crossentropy', optimizer = 'adam')
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 10)                50        
_________________________________________________________________
dense_5 (Dense)              (None, 1)                 11        
_________________________________________________________________
dense_6 (Dense)              (None, 3)                 6         
Total params: 67
Trainable params: 67
Non-trainable params: 0
_________________________________________________________________


In [11]:
# treina o modelo
model.fit(x, y, verbose = 2, epochs = 100)

Epoch 1/100
 - 2s - loss: 1.0998
Epoch 2/100
 - 0s - loss: 1.0994
Epoch 3/100
 - 0s - loss: 1.0994
Epoch 4/100
 - 0s - loss: 1.0989
Epoch 5/100
 - 0s - loss: 1.0986
Epoch 6/100
 - 0s - loss: 1.0985
Epoch 7/100
 - 0s - loss: 1.0984
Epoch 8/100
 - 0s - loss: 1.0982
Epoch 9/100
 - 0s - loss: 1.0984
Epoch 10/100
 - 0s - loss: 1.0980
Epoch 11/100
 - 0s - loss: 1.0980
Epoch 12/100
 - 0s - loss: 1.0979
Epoch 13/100
 - 0s - loss: 1.0979
Epoch 14/100
 - 0s - loss: 1.0976
Epoch 15/100
 - 0s - loss: 1.0975
Epoch 16/100
 - 0s - loss: 1.0974
Epoch 17/100
 - 0s - loss: 1.0973
Epoch 18/100
 - 0s - loss: 1.0971
Epoch 19/100
 - 0s - loss: 1.0970
Epoch 20/100
 - 0s - loss: 1.0969
Epoch 21/100
 - 0s - loss: 1.0967
Epoch 22/100
 - 0s - loss: 1.0965
Epoch 23/100
 - 0s - loss: 1.0963
Epoch 24/100
 - 0s - loss: 1.0960
Epoch 25/100
 - 0s - loss: 1.0957
Epoch 26/100
 - 0s - loss: 1.0956
Epoch 27/100
 - 0s - loss: 1.0951
Epoch 28/100
 - 0s - loss: 1.0948
Epoch 29/100
 - 0s - loss: 1.0942
Epoch 30/100
 - 0s - lo

In [12]:
# Classes encontradas
print(species)

['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']


In [0]:
# Desliga a opção de notação científica para imprimir os resultados.
np.set_printoptions(suppress = True)

Agora que você tem uma rede neural treinada, vamos usá-la. O código a seguir usa nossa rede neural. Exatamente como antes, vamos gerar prédições. Observe que 3 valores voltam para cada uma das 150 flores da íris. Havia 3 tipos de íris (Iris-setosa, Iris-versicolor e Iris-virginica).

In [17]:
np.set_printoptions(suppress=True)
pred = model.predict(x)
print("Shape: {}".format(pred.shape))
print(pred[:10])

Shape: (150, 3)
[[0.87465876 0.02141222 0.10392903]
 [0.8187447  0.03762416 0.14363109]
 [0.8493396  0.02836394 0.12229647]
 [0.8117921  0.03985329 0.1483546 ]
 [0.88179755 0.01957599 0.09862647]
 [0.87492263 0.02134334 0.10373402]
 [0.85051    0.02802798 0.12146204]
 [0.8536035  0.02714668 0.11924984]
 [0.79491717 0.04544919 0.15963362]
 [0.8261748  0.03529222 0.13853298]]


In [16]:
y[:10]

array([[1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.]], dtype=float32)

In [18]:
# Normalmente, a coluna (pred) com maior previsão é considerada a predição da rede neural. 
# A função argmax encontra o índice da previsão máxima para cada linha.
predict_classes = np.argmax(pred,axis=1)
expected_classes = np.argmax(y,axis=1)
print("Valores Previstos: {}".format(predict_classes))
print("Valores Esperados: {}".format(expected_classes))

Valores Previstos: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1]
Valores Esperados: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [19]:
print(species)

['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']


In [20]:
# É muito fácil transformar esses índices de volta em espécies de íris. 
# Nós apenas usamos a lista de espécies que criamos anteriormente.
print(species[predict_classes[1:10]])

['Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa']


In [21]:
# A precisão pode ser uma métrica de erro mais fácil de entender. 
# É essencialmente uma pontuação de teste. Para todas as previsões da íris, que porcentagem estava correta? 
# A desvantagem é que não considera a confiança na rede neural em cada previsão.
from sklearn.metrics import accuracy_score
correct = accuracy_score(expected_classes, predict_classes)
print("Acurácia: {}".format(correct))

Acurácia: 0.66


O código abaixo executa duas previsões ad hoc. A primeira previsão é simplesmente uma única flor de íris. O segundo prediz duas flores de íris. Observe que o argmax na segunda predição requer ** eixo = 1 ** <br > Uma vez que temos um array 2D agora, devemos especificar qual eixo levar o argmax. O valor ** axis = 1 ** especifica que queremos o índice de coluna max para cada linha.

In [22]:
tf.logging.set_verbosity(tf.logging.ERROR)
print(species)

# Previsão ad-hoc
sample_flower = np.array( [[5.0,3.0,4.0,2.0]], dtype=float)
pred = model.predict(sample_flower)
print(pred)
pred = np.argmax(pred)
print("Prevendo que {} é: {}".format(sample_flower,species[pred]))

['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']
[[0.15543687 0.47747836 0.3670848 ]]
Prevendo que [[5. 3. 4. 2.]] é: Iris-versicolor


In [23]:
# predict two sample flowers
sample_flower = np.array( [[5.0,3.0,4.0,2.0],[5.2,3.5,1.5,0.8]], dtype=float)
pred = model.predict(sample_flower)
print(pred)
pred = np.argmax(pred,axis=1)
print("Prevendo que {} é: {}".format(sample_flower,species[pred]))

[[0.15543687 0.47747836 0.3670848 ]
 [0.82423264 0.03589669 0.13987058]]
Prevendo que [[5.  3.  4.  2. ]
 [5.2 3.5 1.5 0.8]] é: ['Iris-versicolor' 'Iris-setosa']
